docs/client-conventions.md

```diff
@@ -0,0 +1,164 @@
+
+Some of the conventions involve the client setting labels or annotations; the
+`client.knative.dev/*` label/annotation namespace is reserved for documented
+Knative client convetions.
+
+## Determine when an action is complete
+
+As Knative is (like all of Kubernetes) a declarative API, the user expresses
+their desire by changing some values in the Knative objects. Clients need not be
+declarative, and might have expressions of user intent like "Deploy this code"
+or "Change these environment variables". To tell when such an action is
+complete, the client can look at the status conditions.
+
+Each Knative object has a `Ready` status condition. When a change is initiated,
+the controller flips this to `Unknown`. When the serving state again reflects
+exactly what the spec of the object specifies, the `Ready` condition will flip
+to `True`; this indicates the operation was a success. If reflecting the spec in
+the serving state is impossible, the `Ready` condition will flip to `False`;
+this indicates the operation was a failure, and the message of the status
+condition should indicate something in English about why (and the Reason field
+can indicate an emumeration suitable for i18n). Either `True` or `False`
+indicates the operation is complete, for better or worse.
+
+Note that someone else could start another operation while the cient was waiting
+for its operation. A conventional client still waits for the `Ready` condition
+to land at `True` or `False`, and then describes to the user what happened using
+logic based on the intended effect.
+
+For example:
+ * Client A deploys image `gcr.io/foods/vegetables:eggplant`
+ * While that is not yet Ready, client B deploys `gcr.io/foods/vegetables:squash`
+ * The `eggplant` revision becomes Ready: True, and the service moves traffic to it.
+ * The `squash` revision fails to bind to a port, and becomes Ready: False
+ * The Service switches from Ready: Unknown to Ready: False because `squash` failed.
+
+Both client A and B should wait for the last step in this procedure.
+ * Client A sees that latestReadyRevision is the revision with the
+   [nonce](#associate-modifications-with-revisions) it specified, and that
+   `latestCreatedRevision` is not. It tells the user that deploying was
+   successful.
+ * Client B sees that `latestCreatedRevision` is the revision with the nonce it
+   specified; it reports the failure with the appropriate message.
+
+The rule is "Wait for `Ready` to become `True` or `False`, then report on
+whether your intent was accomplished". The `Ready` success or failure can be
+part of this report, but may be confusing (as in the example) if it's the only
+thing you report.
+
+## Associate modifications with Revisions
+
+Every time the client changes a Service or Configuration in a way that results
+in a new Revision, it should change the label `client.knative.dev/nonce` to a
+new random value, chosen from a wide enough space to avoid collisions.
```

Comment:
_(sorry for not thinking of this earlier, but fresh eyes)_

tl;dr once we can rely on `metadata.generation` the client should be able to use that for the lookups you are using the nonce for.

Basically: `metadata.generation` (but for now `spec.generation`) is synchronously updated with requests to the next sequence number reflecting the user's edits.  We label each revision with `serving.knative.dev/configurationGeneration: NNN`.  So the client would take it's receipt of the update and use that to construct a label selector.

This feels like a much stronger (system-enforced) way of looking up the resulting Revision, and means the `nonce` need only change when the user wants to **force** an update.

We should be able to move this direction in the 0.3 timeframe (I hope).

Reply:
I see you mention the `.../configuration` label in the next section, so perhaps we discussed this before and there were good reasons.

Reply:
Oh, I think the problem here is the extra layer of Steve.  I need more coffee.

Reply:
Yeah, the problem here is the extra layer of Service. Also I am a little unclear on how guaranteed `generation` is to be the next ordinal when you modify stuff. The k8s documetation for developers says:

> generation: a sequence number representing a specific generation of the desired state. Set by the system and monotonically increasing, per-resource. May be compared, such as for RAW and WAW consistency.

This doesn't make me feel like I can anticipate it, even if in practice I can?

Reply:
You shouldn't need to anticipate it, right?  You'd be able to say:
```
cfg, err := configs.Create(cfg)
// handle err

// Use the generation synchronously updated with the request.
selector := fmt.Sprintf("serving.knative.dev/Configuration=%d", cfg.metadata)
```

I believe this would work, if we only had Configuration, but we also have Service. :(

Reply:
I guess revisions could have a service-generation label too?

Reply:
I believe that would also work.